### PR TITLE
Fix Issue #22140: FunctionTypeOf fails for template opCall without any templated argument

### PR DESCRIPTION
**Pull request by ljmf00**
Signed-off-by: Luís Ferreira <contact@lsferreira.net>

### CODE REVIEW

File: std/traits.d [hunk 1]
```diff
@@ -2279,15 +2279,18 @@ usually used for compile-time reflection purposes.
 template FunctionTypeOf(func...)
 if (func.length == 1 && isCallable!func)
 {
-    static if (is(typeof(& func[0]) Fsym : Fsym*) && is(Fsym == function) || is(typeof(& func[0]) Fsym == delegate))
+    static if ((is(typeof(& func[0]) Fsym : Fsym*) && is(Fsym == function)) || is(typeof(& func[0]) Fsym == delegate))
     {
         alias FunctionTypeOf = Fsym; // HIT: (nested) function symbol
     }
-    else static if (is(typeof(& func[0].opCall) Fobj == delegate))
+    else static if (is(typeof(& func[0].opCall) Fobj == delegate) || is(typeof(& func[0].opCall!()) Fobj == delegate))
```
Comment (ljmf00) [inline std/traits.d:2286]: Can someone provide me with an example of where this is applicable?

File: std/traits.d [hunk 2]
```diff
@@ -2279,15 +2279,18 @@ usually used for compile-time reflection purposes.
 template FunctionTypeOf(func...)
 if (func.length == 1 && isCallable!func)
 {
-    static if (is(typeof(& func[0]) Fsym : Fsym*) && is(Fsym == function) || is(typeof(& func[0]) Fsym == delegate))
+    static if ((is(typeof(& func[0]) Fsym : Fsym*) && is(Fsym == function)) || is(typeof(& func[0]) Fsym == delegate))
```
Comment (RazvanN7) [inline std/traits.d:2282]: This change is unnecessary

**Review by RazvanN7 — CHANGES_REQUESTED** [review]
Needs tests.

File: std/traits.d (hunk 1 above)
Comment (RazvanN7) [reply std/traits.d:2286]:
```d
void fun()
{
    int a;
    struct B { int opCall() { return a; }}
}
```
}

**Comment by RazvanN7** [comment]
ping @ljmf00

File: std/traits.d (hunk 2 above)
Comment (ljmf00) [reply std/traits.d:2282]: It is necessary for clarity. I based my change on DScanner suggestions. If you prefer I can revert it, but the rationale is that it clarifies the logic precedence of the expression.

https://github.com/dlang-community/D-Scanner/blob/2963358eb4a24064b0893493684d4075361297eb/src/dscanner/analysis/logic_precedence.d#L45

File: std/traits.d (hunk 1 above)
Comment (ljmf00) [reply std/traits.d:2286]: Done

**Comment by ljmf00** [comment]
> ping @ljmf00

Thanks for the ping :+1:

**Review by RazvanN7 — APPROVED** [review]

**Comment by RazvanN7** [comment]
Thanks @ljmf00 !

**Pull request merged**